website/docs/tutorials/nextjs.md

```diff
@@ -0,0 +1,313 @@
+
+```tsx
+// src/app/[locale]/layout.tsx
+import { t } from '@lingui/macro'
+import I18nProvider from '@/components/I18nProvider'
+import { loadCatalog, setI18n } from '@/utils/locales'
+
+type Params = {
+  locale: string
+}
+
+type Props = {
+  params: Params
+  children: React.ReactNode
+}
+
+export function generateMetadata({ params }: { params: Params }) {
+  const { locale } = params
+  const i18n = setI18n(locale)
+  return {
+    title: t(i18n)`Translation Demo`, // this is an example of a server only translation
+  }
+}
+
+export function generateStaticParams() {
+    return [{ locale: 'en' }, { locale: 'fr' }]
+}
+
+export default function RootLayout({ params, children }: Props) {
+    const { locale } = params
+    const messages = loadCatalog(locale)
+    setI18n(locale)
+    return (
+        <html lang={locale}>
+            <body>
+                <I18nProvider locale={locale} messages={messages}>
+                    {children}
+                </I18nProvider>
+            </body>
+        </html>
+    )
+}
+```
+
+You can use the Next.js middleware to [redirect to the default language](https://nextjs.org/docs/app/building-your-application/routing/internationalization) if the root `/` is visited.
+
+```ts
+// src/middleware.ts
+import { NextResponse, NextRequest } from "next/server"
+ 
+const locales = ['en', 'fr']
+ 
+export function middleware(req : NextRequest) {
+  const { pathname } = req.nextUrl
+  const pathnameHasLocale = locales.some(
+    (locale) => pathname.startsWith(`/${locale}/`) || pathname === `/${locale}`
+  )
+  if (pathnameHasLocale) return
+ 
+  const defaultLocale = 'en' // or find the "accept-language" in the header
+  req.nextUrl.pathname = `/${defaultLocale}${pathname}`
+  return NextResponse.redirect(req.nextUrl)
+}
+ 
+export const config = {
+  matcher: [
+    '/((?!_next).*)',
+  ],
+}
+```
+
+As you can see in the example above we slight modified the default `I18nProvider` with our own implementation. Here we pass the active message catalog as a prop. This is to avoid importing all catalogs inside the `I18nProvider` component itself and thus making them part of the client bundle. For the people who are wondering: Why we are not just passing an instance of `i18n` the original provider component? Server components can pass data to client components but the caveat is that this needs to be serializable data and `i18n` is not.
+
+```tsx
+// src/components/I18nProvider.ts
+'use client'
+
+import { setupI18n } from '@lingui/core'
+import { I18nProvider as LinguiProvider } from '@lingui/react'
+
+type Props = {
+  locale: string
+  messages?: any
+  children: React.ReactNode
+}
+
+export function I18nProvider({ locale, messages, ...props }: Props) {
```

Comment:
Would  it be better to use a `useMemo` with `[locale, messages]` deps? 

I don't really understand why it's became `initial` since, when language switched this provider would be called with new props